// zcatalyst-sdk-node/lib/job-scheduling/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JOB_SOURCE_TYPE = exports.JOB_STATUS = exports.TARGET_TYPE = exports.CRON_TYPE = exports.CRON_EXECUTION_TYPE = exports.CAPACITY_ATTRIBUTE_TYPE = void 0;
/**
 * Attributes that are used to determine the capacity of the jobpool
 */
var CAPACITY_ATTRIBUTE_TYPE;
(function (CAPACITY_ATTRIBUTE_TYPE) {
    CAPACITY_ATTRIBUTE_TYPE["MEMORY"] = "memory";
    CAPACITY_ATTRIBUTE_TYPE["NUMBER"] = "number";
})(CAPACITY_ATTRIBUTE_TYPE = exports.CAPACITY_ATTRIBUTE_TYPE || (exports.CAPACITY_ATTRIBUTE_TYPE = {}));
/**
 * Execution type of the cron
 */
var CRON_EXECUTION_TYPE;
(function (CRON_EXECUTION_TYPE) {
    /** Static crons that are created from catalyst console */
    CRON_EXECUTION_TYPE["PRE_DEFINED"] = "pre-defined";
    /** Dynamic crons that are created programmatically */
    CRON_EXECUTION_TYPE["DYNAMIC"] = "dynamic";
})(CRON_EXECUTION_TYPE = exports.CRON_EXECUTION_TYPE || (exports.CRON_EXECUTION_TYPE = {}));
/**
 * Type of the cron
 */
var CRON_TYPE;
(function (CRON_TYPE) {
    /** Repetitive cron that's executed during a calendar time */
    CRON_TYPE["CALENDER"] = "Calender";
    /** Repetitive cron that's executed in a particular period of time */
    CRON_TYPE["PERIODIC"] = "Periodic";
    /** Cron that's executed only once on the scheduled time */
    CRON_TYPE["ONETIME"] = "OneTime";
    /** Cron's that are created using the UNIX cron expression */
    CRON_TYPE["CRON_EXPRESSION"] = "CronExpression";
})(CRON_TYPE = exports.CRON_TYPE || (exports.CRON_TYPE = {}));
/**
 * Types of job targets
 */
var TARGET_TYPE;
(function (TARGET_TYPE) {
    TARGET_TYPE["FUNCTION"] = "Function";
    TARGET_TYPE["CIRCUIT"] = "Circuit";
    TARGET_TYPE["APP_SAIL"] = "AppSail";
    TARGET_TYPE["WEBHOOK"] = "Webhook";
})(TARGET_TYPE = exports.TARGET_TYPE || (exports.TARGET_TYPE = {}));
/**
 * Status of the job
 */
var JOB_STATUS;
(function (JOB_STATUS) {
    /** Submitted to the job pool */
    JOB_STATUS["SUBMITTED"] = "Submitted";
    /** Awaiting execution */
    JOB_STATUS["PENDING"] = "Pending";
    /** Job being executed */
    JOB_STATUS["RUNNING"] = "Running";
    /** Job executed successfully */
    JOB_STATUS["SUCCESSFUL"] = "Successful";
    /** Job execution failed */
    JOB_STATUS["FAILURE"] = "Failure";
})(JOB_STATUS = exports.JOB_STATUS || (exports.JOB_STATUS = {}));
/**
 * Source from which the job is submitted
 */
var JOB_SOURCE_TYPE;
(function (JOB_SOURCE_TYPE) {
    JOB_SOURCE_TYPE["API"] = "API";
    JOB_SOURCE_TYPE["CRON"] = "Cron";
})(JOB_SOURCE_TYPE = exports.JOB_SOURCE_TYPE || (exports.JOB_SOURCE_TYPE = {}));
