// zcatalyst-sdk-node/lib/smartbrowz/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Smartbrowz_instances, _Smartbrowz_baseUri, _Smartbrowz_dataverse, _Smartbrowz_browserGrid, _Smartbrowz_execute;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Smartbrowz = void 0;
const constants_1 = __importDefault(require("../utils/constants"));
const api_request_1 = require("../utils/api-request");
const validator_1 = require("../utils/validator");
const error_1 = require("../utils/error");
const dataverse_1 = require("./dataverse");
const browser_grid_1 = require("./browser-grid");
const { REQ_METHOD, CREDENTIAL_USER, PRODUCT_NAME, API_VERSION } = constants_1.default;
class Smartbrowz {
    constructor(app) {
        _Smartbrowz_instances.add(this);
        _Smartbrowz_baseUri.set(this, void 0);
        _Smartbrowz_dataverse.set(this, void 0);
        _Smartbrowz_browserGrid.set(this, void 0);
        (0, validator_1.isValidApp)(app, true);
        this.projectId = app.config.projectId;
        this.requester = new api_request_1.AuthorizedHttpClient(app);
        __classPrivateFieldSet(this, _Smartbrowz_baseUri, `/${PRODUCT_NAME.browser360}/${API_VERSION}/project/${this.projectId}`, "f");
        __classPrivateFieldSet(this, _Smartbrowz_dataverse, new dataverse_1.Dataverse({ baseUri: __classPrivateFieldGet(this, _Smartbrowz_baseUri, "f"), requester: this.requester }), "f");
        __classPrivateFieldSet(this, _Smartbrowz_browserGrid, new browser_grid_1.BrowserGrid({ baseUri: __classPrivateFieldGet(this, _Smartbrowz_baseUri, "f"), requester: this.requester }), "f");
    }
    /**
     * Access a Dataverse instance
     * @returns an instance object to access Dataverse functionalities
     */
    dataverse() {
        return __classPrivateFieldGet(this, _Smartbrowz_dataverse, "f");
    }
    /**
     * Access a BrowserGrid instance
     * @returns an instance object to access Browser Grid functionalities
     */
    browserGrid() {
        return __classPrivateFieldGet(this, _Smartbrowz_browserGrid, "f");
    }
    convertToPdf(source, options) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(source, 'url or html', true);
            }, error_1.CatalystSmartbrowzError);
            const pdfOptions = Object.assign(Object.assign(Object.assign({}, ((0, validator_1.isURL)(source) ? { url: source } : { html: source })), { output_options: {
                    output_type: 'pdf'
                } }), options);
            return yield __classPrivateFieldGet(this, _Smartbrowz_instances, "m", _Smartbrowz_execute).call(this, pdfOptions);
        });
    }
    takeScreenshot(source, options) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(source, 'url or html', true);
            }, error_1.CatalystSmartbrowzError);
            const ScrOptions = Object.assign(Object.assign(Object.assign({}, ((0, validator_1.isURL)(source) ? { url: source } : { html: source })), { output_options: {
                    output_type: 'screenshot'
                } }), options);
            return yield __classPrivateFieldGet(this, _Smartbrowz_instances, "m", _Smartbrowz_execute).call(this, ScrOptions);
        });
    }
    generateFromTemplate(id, options) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyStringOrNumber)(id, 'template id', true);
            }, error_1.CatalystSmartbrowzError);
            const templateOptions = Object.assign({ template_id: id }, options);
            return yield __classPrivateFieldGet(this, _Smartbrowz_instances, "m", _Smartbrowz_execute).call(this, templateOptions);
        });
    }
    /**
     * Get comprehensive details about any organization using its name, email address or website URL.
     * @param leadDetails - detail to identify the lead. Supported values Email, LeadName, WebsiteURL.
     * @returns enriched leads data
     *
     * @deprecated This method is deprecated and will be removed in the future version. \
     * Use {@link Dataverse.getEnrichedLead getEnrichedLead} method instead
     *
     * @example
     * ```
     * // AdvancedIO
     * const catalyst = require('zcatalyst-sdk-node');
     * const app = catalyst.initialize(req); // initialize catalyst sdk
     * await app.SmartBrowz().dataverse().getEnrichedLead({websiteUrl: "https://catalyst.zoho.com"}); // get Enriched Leads
     * await app.SmartBrowz().getEnrichedLead({websiteUrl: "https://catalyst.zoho.com"}); // get Enriched Leads - deprecated
     * ```
     */
    getEnrichedLead({ email, leadName, websiteUrl } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line no-console
            console.warn('This method "getEnrichedLead" is deprecated and will be removed in a future version');
            return __classPrivateFieldGet(this, _Smartbrowz_dataverse, "f").getEnrichedLead({ email, leadName, websiteUrl });
        });
    }
    /**
     * Get details about the technologies and frameworks used by an organization.
     * @param websiteUrl - Url of the website to find the tech stack.
     * @returns tech stack details of the website
     *
     * @deprecated This method is deprecated and will be removed in the future version. \
     * Use {@link Dataverse.findTechStack findTechStack} method instead
     *
     * @example
     * ```
     * // AdvancedIO
     * const catalyst = require('zcatalyst-sdk-node');
     * const app = catalyst.initialize(req); // initialize catalyst sdk
     * await app.SmartBrowz().dataverse().findTechStack("https://catalyst.zoho.com"); // find tech stack
     * await app.SmartBrowz().findTechStack("https://catalyst.zoho.com"); // find tech stack - deprecated
     * ```
     */
    findTechStack(websiteUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line no-console
            console.warn('This method "findTechStack" is deprecated and will be removed in a future version');
            return __classPrivateFieldGet(this, _Smartbrowz_dataverse, "f").findTechStack(websiteUrl);
        });
    }
    /**
     * Find out all the potential competitors of an organization.
     * @param orgDetails - details to identify the organization. Supported values are LeadName and WebsiteURL
     * @returns list of similar organizations.
     *
     * @deprecated This method is deprecated and will be removed in the future version. \
     * Use {@link Dataverse.getSimilarCompanies getSimilarCompanies} method instead
     *
     * @example
     * ```
     * // AdvancedIO
     * const catalyst = require('zcatalyst-sdk-node');
     * const app = catalyst.initialize(req); // initialize catalyst sdk
     * await app.SmartBrowz().dataverse().getSimilarCompanies({ leadName: "zoho catalyst" }); // find tech stack
     * await app.SmartBrowz().getSimilarCompanies({ leadName: "zoho catalyst"}); // find tech stack - deprecated
     */
    getSimilarCompanies({ leadName, websiteUrl } = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line no-console
            console.warn('This method "getSimilarCompanies" is deprecated and will be removed in a future version');
            return __classPrivateFieldGet(this, _Smartbrowz_dataverse, "f").getSimilarCompanies({ leadName, websiteUrl });
        });
    }
}
exports.Smartbrowz = Smartbrowz;
_Smartbrowz_baseUri = new WeakMap(), _Smartbrowz_dataverse = new WeakMap(), _Smartbrowz_browserGrid = new WeakMap(), _Smartbrowz_instances = new WeakSet(), _Smartbrowz_execute = function _Smartbrowz_execute(details) {
    return __awaiter(this, void 0, void 0, function* () {
        const request = {
            method: REQ_METHOD.post,
            path: __classPrivateFieldGet(this, _Smartbrowz_baseUri, "f") + '/convert',
            type: "json" /* RequestType.JSON */,
            expecting: "raw" /* ResponseType.RAW */,
            data: details,
            catalyst: false,
            track: true,
            user: CREDENTIAL_USER.admin
        };
        const resp = yield this.requester.send(request);
        return resp.data;
    });
};
