// zcatalyst-sdk-node/lib/connections/connections.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connections = void 0;
const validator_1 = require("../utils/validator");
const api_request_1 = require("../utils/api-request");
const error_1 = require("../utils/error");
const constants_1 = __importDefault(require("../utils/constants"));
const { REQ_METHOD, CREDENTIAL_USER, COMPONENT } = constants_1.default;
class Connections {
    constructor(app) {
        (0, validator_1.isValidApp)(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
    }
    getComponentName() {
        return COMPONENT.connections;
    }
    /**
     * Retrieves connection credentials for a specific connection.
     * @param {string} connectionLinkName - The unique name of the connection
     * @returns {ICatalystConnectionsResponse} A promise that resolves with the connection credentials
     * @throws {CatalystConnectionsError} If the connectionLinkName is invalid or connection retrieval fails
     */
    getConnectionCredentials(connectionLinkName) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(connectionLinkName, 'connectionLinkName', true);
            }, error_1.CatalystConnectionsError);
            const request = {
                method: REQ_METHOD.get,
                path: '/connection-details',
                type: "json" /* RequestType.JSON */,
                expecting: "json" /* ResponseType.JSON */,
                qs: { 'connection-link-name': connectionLinkName },
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const response = yield this.requester.send(request);
            return response.data.data;
        });
    }
}
exports.Connections = Connections;
