// zcatalyst-sdk-node/lib/pipeline/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pipeline = void 0;
const api_request_1 = require("../utils/api-request");
const constants_1 = __importDefault(require("../utils/constants"));
const error_1 = require("../utils/error");
const validator_1 = require("../utils/validator");
const { CREDENTIAL_USER, REQ_METHOD, COMPONENT } = constants_1.default;
class Pipeline {
    constructor(app) {
        (0, validator_1.isValidApp)(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
    }
    getComponentName() {
        return COMPONENT.pipeline;
    }
    /**
     * To get the pipeline details
     * @param pipelineId - Id of the pipeline
     * @returns Pipeline details
     */
    getPipelineDetails(pipelineId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(pipelineId, 'pipeline id', true);
            }, error_1.CatalystPipelineError);
            const request = {
                method: REQ_METHOD.get,
                path: `/pipeline/${pipelineId}`,
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    /**
     * Trigger the pipeline by Id
     * @param pipelineId - Id of the pipeline
     * @param branch - branch name in the repo
     * @param envVariables - Environment variables
     * @returns Pipeline run response
     */
    runPipeline(pipelineId, branch, envVariables) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(pipelineId, 'pipeline id', true);
            }, error_1.CatalystPipelineError);
            const request = {
                method: REQ_METHOD.post,
                path: `/pipeline/${pipelineId}/run`,
                qs: branch ? { branch } : {},
                data: envVariables || {},
                catalyst: true,
                type: "json" /* RequestType.JSON */,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
}
exports.Pipeline = Pipeline;
