// zcatalyst-sdk-node/lib/job-scheduling/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _JobScheduling_cron, _JobScheduling_job;
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobScheduling = void 0;
const api_request_1 = require("../utils/api-request");
const validator_1 = require("../utils/validator");
const constants_1 = __importDefault(require("../utils/constants"));
const constants_2 = __importDefault(require("./constants"));
const job_1 = __importDefault(require("./job"));
const error_1 = __importDefault(require("./error"));
const cron_1 = __importDefault(require("./cron"));
const { COMPONENT, REQ_METHOD, CREDENTIAL_USER } = constants_1.default;
const { API } = constants_2.default;
class JobScheduling {
    constructor(app) {
        _JobScheduling_cron.set(this, void 0);
        _JobScheduling_job.set(this, void 0);
        (0, validator_1.isValidApp)(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
        __classPrivateFieldSet(this, _JobScheduling_cron, new cron_1.default(this.requester), "f");
        __classPrivateFieldSet(this, _JobScheduling_job, new job_1.default(this.requester), "f");
    }
    getComponentName() {
        return COMPONENT.job_scheduling;
    }
    /**
     * Property to access Cron functionalities
     *
     * @deprecated - This property is deprecated and will be removed in the future version. \
     * Use {@link JobScheduling.cron() cron()} method instead.
     */
    get CRON() {
        // eslint-disable-next-line no-console
        console.warn('This property "CRON" is deprecated and will be removed in a future version.');
        return __classPrivateFieldGet(this, _JobScheduling_cron, "f");
    }
    /**
     * Property to access Job functionalities
     *
     * @deprecated - This property is deprecated and will be removed in the future version. \
     * Use {@link JobScheduling.job() job()} method instead.
     */
    get JOB() {
        // eslint-disable-next-line no-console
        console.warn('This property "JOB" is deprecated and will be removed in a future version.');
        return __classPrivateFieldGet(this, _JobScheduling_job, "f");
    }
    /**
     * Get an object instance to access Cron functionalities
     * @returns Cron instance
     */
    cron() {
        return __classPrivateFieldGet(this, _JobScheduling_cron, "f");
    }
    /**
     * Get an object instance to access Job functionalities
     * @returns Job instance
     */
    job() {
        return __classPrivateFieldGet(this, _JobScheduling_job, "f");
    }
    // Jobpool
    /**
     * Get Jobpool details of the specific Jobpool
     * @param jobpoolId - Id of the Jobpool
     * @returns Jobpool details
     */
    getJobpool(jobpoolId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyStringOrNumber)(jobpoolId, 'jobpool_id', true);
            }, error_1.default);
            const request = {
                method: REQ_METHOD.get,
                path: API.JOBPOOL + '/' + jobpoolId,
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
    /**
     * Get all jobpool details
     * @returns Array of jobpool details
     */
    getAllJobpool() {
        return __awaiter(this, void 0, void 0, function* () {
            const request = {
                method: REQ_METHOD.get,
                path: API.JOBPOOL,
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
}
exports.JobScheduling = JobScheduling;
_JobScheduling_cron = new WeakMap(), _JobScheduling_job = new WeakMap();
