// zcatalyst-sdk-node/lib/smartbrowz/browser-grid.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _BrowserGrid_requester, _BrowserGrid_baseUri;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserGrid = void 0;
const constants_1 = __importDefault(require("../utils/constants"));
const { REQ_METHOD, CREDENTIAL_USER } = constants_1.default;
/**
 * `BrowserGrid`
 * Class that contains the methods to access Catalyst SmartBrowz functionalities
 */
class BrowserGrid {
    constructor({ requester, baseUri }) {
        _BrowserGrid_requester.set(this, void 0);
        _BrowserGrid_baseUri.set(this, void 0);
        __classPrivateFieldSet(this, _BrowserGrid_requester, requester, "f");
        __classPrivateFieldSet(this, _BrowserGrid_baseUri, baseUri + '/browser-grid', "f");
    }
    getGrid(gridId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = {
                method: REQ_METHOD.get,
                path: `${__classPrivateFieldGet(this, _BrowserGrid_baseUri, "f")}${gridId ? '/' + gridId : ''}`,
                expecting: "json" /* ResponseType.JSON */,
                catalyst: false,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield __classPrivateFieldGet(this, _BrowserGrid_requester, "f").send(request);
            return resp.data.data;
        });
    }
    /**
     * Get the data on Nodes in a Browser Grid
     * @param gridId Id or Name of the Browser Grid
     * @returns Data on the Nodes of the Browser Grid
     *
     * @example
     * ```
     * // AdvancedIO
     *
     * // import catalyst sdk
     * const catalyst = require('zcatalyst-sdk-node');
     *
     * // initialize catalyst sdk
     * const app = catalyst.initialize(req);
     *
     * // get browser grid details with grid name
     * const nodes = await app.SmartBrowz()
     * 					.browserGrid()
     * 					.getGridNodes("testGrid");
     */
    getGridNodes(gridId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = {
                method: REQ_METHOD.get,
                path: `${__classPrivateFieldGet(this, _BrowserGrid_baseUri, "f")}/${gridId}/stats?data_to_fetch=live_stats`,
                expecting: "json" /* ResponseType.JSON */,
                catalyst: false,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield __classPrivateFieldGet(this, _BrowserGrid_requester, "f").send(request);
            return resp.data.data;
        });
    }
    /**
     * Stop the Nodes in a Browser Grid
     * @param gridId Id or Name of the Browser Grid
     *
     * @example
     * ```
     * // AdvancedIO
     *
     * // import catalyst sdk
     * const catalyst = require('zcatalyst-sdk-node');
     *
     * // initialize catalyst sdk
     * const app = catalyst.initialize(req);
     *
     * // get browser grid details with grid name
     * await app.SmartBrowz().browserGrid().stopGrid("testGrid");
     */
    stopGrid(gridId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = {
                method: REQ_METHOD.post,
                path: `${__classPrivateFieldGet(this, _BrowserGrid_baseUri, "f")}/${gridId}/stop`,
                expecting: "json" /* ResponseType.JSON */,
                catalyst: false,
                track: true,
                user: CREDENTIAL_USER.admin
            };
            const resp = yield __classPrivateFieldGet(this, _BrowserGrid_requester, "f").send(request);
            return resp.data.data;
        });
    }
}
exports.BrowserGrid = BrowserGrid;
_BrowserGrid_requester = new WeakMap(), _BrowserGrid_baseUri = new WeakMap();
